src/librustc_codegen_ssa/back/linker.rs

```diff
@@ -368,6 +368,23 @@ impl<'a> Linker for GccLinker<'a> {
             }
         } else {
             self.cmd.arg("-shared");
+            if self.sess.target.target.options.is_like_windows {
+                let implib_name = out_filename
+                    .file_name()
+                    .and_then(|file| file.to_str())
+                    .map(|file| format!("{}{}{}",
+                         self.sess.target.target.options.staticlib_prefix,
+                         file,
+                         self.sess.target.target.options.staticlib_suffix));
```

Comment:
This looks like it's generating a file called `libfoo.a`, but is it intended that this is generating `libfoo.dll.a`?

Also does this already happen on the existing MinGW target implicitly without this argument being generated? Or otherwise?

If it's not already happening on MinGW, can this be gated to only happen for UWP?

Reply:
It kind of happens implicitly with GCC/mingw as it can link directly with a DLL, while Clang can't  

To be fair I'm not sure how the GNU linker handles it. My understanding is that MSVC doesn't support directly linking to a DLL and requires an import library, and clang is aiming at a feature parity with MSVC more than GCC when it comes to building for Windows.

Though that really is my sole understanding only.

In any case, it already works as is with GCC, but it also works with the implib being explicitly generated. If you prefer so, I can gate it for UWP.

Regarding the naming convention, I can definitely make it more like `libfoo.dll.a`, I'm not sure there's a clear and definite convention, all the import libraries from mingw are in the form libfoo.a to link with foo.dll, I think MSVC aims at foo.lib, and I've indeed seen some .dll.a

Reply:
I think it's probably best to just go ahead and align with what MSVC is doing. I don't really understand why the MinGW UWP toolchain uses LLVM and departs from MinGW idioms, or why a MSVC-based UWP toolchain doesn't work, but we may as well try to keep things at least a little consistent on Windows as we can try...

And yes I think this block should be UWP specific because it's not needed on the MinGW toolchain.

Reply:
According to https://github.com/rust-lang/rust/pull/29520 we emit `foo.dll.lib` as the import library for `-msvc` so to emit `libfoo.dll.a` for `-gnu` makes perfect sense for me. While MinGW is odd in that it can to some extent link to dlls without an import library, I'm in favor of always emitting import libraries on Windows regardless of the toolchain.